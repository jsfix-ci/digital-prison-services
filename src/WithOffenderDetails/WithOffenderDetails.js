import React from 'react'
import PropTypes from 'prop-types'
import axios from 'axios'
import { connect } from 'react-redux'
import { resetError, setLoaded, setOffender } from '../redux/actions'

export class WithOffenderDetails extends React.Component {
  async componentDidMount() {
    const {
      setLoaded: setLoadedDispatch,
      resetError: resetErrorDispatch,
      setOffender: setOffenderDispatch,
      offenderNumber,
      handleError,
      client,
    } = this.props

    resetErrorDispatch()
    try {
      const { data } = await client.get(`/api/offenders/${offenderNumber}`)
      setLoadedDispatch(true)
      setOffenderDispatch(data)
    } catch (error) {
      handleError(error)
    }
  }

  render() {
    const { children } = this.props
    return children
  }
}

WithOffenderDetails.propTypes = {
  client: PropTypes.func,
  offenderNumber: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  handleError: PropTypes.func.isRequired,
  /* Map dispatch to props */
  setLoaded: PropTypes.func.isRequired,
  setOffender: PropTypes.func.isRequired,
  resetError: PropTypes.func.isRequired,
}

WithOffenderDetails.defaultProps = {
  client: axios,
}

const mapStateToProps = () => ({})

const mapDispatchToProps = {
  setLoaded,
  resetError,
  setOffender,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WithOffenderDetails)
