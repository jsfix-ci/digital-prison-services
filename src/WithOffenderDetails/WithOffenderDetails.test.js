import React from 'react'
import { shallow } from 'enzyme'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import testRenderer from 'react-test-renderer'

import { WithOffenderDetails } from './WithOffenderDetails'

describe('<WithOffenderDetails />', () => {
  const waitForAsync = () => new Promise(resolve => setImmediate(resolve))

  const offender = { firstName: 'Harry', lastName: 'Smith' }

  let mock
  let props
  let setOffender
  let setLoaded
  let handleError
  let resetError

  beforeEach(() => {
    setOffender = jest.fn()
    setLoaded = jest.fn()
    handleError = jest.fn()
    resetError = jest.fn()
    props = { offenderNumber: 'OFF-1', setLoaded, setOffender, handleError, resetError, client: axios }

    mock = new MockAdapter(axios)
  })

  describe('with all props', () => {
    it('On successful api call it populates offender details', async () => {
      mock.onGet(`/api/offenders/${props.offenderNumber}`).reply(200, offender)

      shallow(
        <WithOffenderDetails {...props}>
          <h1>Hello</h1>
        </WithOffenderDetails>
      )
      await waitForAsync()

      expect(resetError).toBeCalled()
      expect(setLoaded).toBeCalled()
      expect(setOffender).toBeCalledWith(offender)
      expect(handleError).not.toBeCalled()
    })

    it('On api call failure it handles the error', async () => {
      mock.onGet(`/api/offenders/${props.offenderNumber}`).reply(500, offender)

      shallow(
        <WithOffenderDetails {...props}>
          <h1>Hello</h1>
        </WithOffenderDetails>
      )
      await waitForAsync()

      expect(resetError).toBeCalled()
      expect(setLoaded).not.toBeCalled()
      expect(setOffender).not.toBeCalled()
      expect(handleError).toBeCalledWith(Error('Request failed with status code 500'))
    })

    it('it renders children', async () => {
      const element = testRenderer.create(
        <WithOffenderDetails {...props}>
          <h1>Hello</h1>
        </WithOffenderDetails>
      )
      await waitForAsync()

      expect(element).toMatchSnapshot()
    })
  })
})
