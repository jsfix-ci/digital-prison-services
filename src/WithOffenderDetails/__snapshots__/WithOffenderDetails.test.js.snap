// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WithOffenderDetails /> with all props it renders children 1`] = `
<h1>
  Hello
</h1>
`;
